### ISSUE TITLE
Android: unable to authenticate

### ISSUE THREAD

**Issue by maxhawkins** [tailscale/tailscale#470]
**Describe the bug**
Nothing happens when I press the Sign In button.

**To Reproduce**
1. Uninstall app
2. Reinstall app
3. Press sign in button
4. Nothing happens

**Screenshots**
![Screenshot_20200617-175953](https://user-images.githubusercontent.com/28565/84965802-bc73b680-b0c4-11ea-8569-ca28f4225953.png)

**Version information:**
 - Device: Pixel 3a
 - OS: Android
 - OS version: 10
 - Tailscale version: 0.1.7

**Additional context**
I uninstalled and reinstalled because on 0.1.7 the VPN didnt disable when I pressed the toggle. I hoped starting over might solve the issue. Here's what that looked like:

![Screenshot_20200617-173857~2](https://user-images.githubusercontent.com/28565/84966088-710dd800-b0c5-11ea-8f9b-7a9830539b54.png)

**Comment by bradfitz** [comment]
It works fine on my Pixel 3a, so we need to figure out what's different between your Pixel and mine.

Mine's pretty stock. I factory reset it recently after loaning it out to somebody.

@maxhawkins, @ericmigi: can you think of anything that might be unusual about your devices? What type of accounts? Other existing VPNs? Which browser?

But I'll let @eliasnaur lead the debugging on this, as he'll probably know what to look at, and where.

**Comment by maxhawkins** [comment]
I updated Android System Webview to 83.0.4103.106 at the same time I updated Tailscale. That could be related.

However, before I reinstalled 0.1.7 from scratch to try and fix the toggle issue I was able to successfully log out and back in using the webview. The problem only happens for me on the initial sign in screen.

**Comment by ericmigi** [comment]
I am on android beta 11

**Comment by jondlove** [comment]
I am having the same issue, here's a screen recording:
==> https://www.dropbox.com/s/1keot0lwnpq3x2u/2020-06-18-14-31-56_1.mp4?dl=0

**Steps to Reproduce:**
- Install 'Tailscale app (Early Access)'
- Open Tailscale app
- Press 'Sign In'

Freshly rebooted? Yes
Freshly installed? Yes (before reboot)

**Version Info:**
Device: Zenfone 6 (ZS630KL)
OS: Android 10
Tailscale: 0.1.7
Play Store: 20.4.18-all
Existing VPNs? Yes, managed by "OpenVPN Client". Not connected while testing.

**Comment by idealisms** [comment]
I have the same bug on a Pixel 3. Here are all my accounts + reddit is fun off screen.

I also have OpenVPN installed and wasn't using it at the time.

![Screenshot_20200618-211642~2](https://user-images.githubusercontent.com/175378/85096560-79d7da00-b1a9-11ea-896b-2d3222fce760.png)

**Comment by maxhawkins** [comment]
I do not have openvpn or any other vpn installed, fwiw.

**Comment by milanaleksic** [comment]
I am on Samsung S10, Android 10 with the same issue. No VPNs active at the time, but OpenVPN and PIA VPN applications installed. I can't go further than the button to Sign In, as in the above screenshot.

**Comment by LvanWissen** [comment]
I have the same issue on my Sony Xperia XZ1 Compact, running Android 9.

**Comment by dishanrajapaksha** [comment]
Same issue in my Galaxy Note 9 with Android 10.

**Comment by xan-it** [comment]
Same issue on Samsung S10, Android 10 with no installed VPN clients nor connections

**Comment by bradfitz** [comment]
@eliasnaur, perhaps `peer jni.Object` is uninitialized and thus nil when `case <-a.updates:`  fires in `(*App).runUI`, which calls:

```go
func (a *App) updateState(javaPeer jni.Object, state *clientState) {
        if javaPeer != 0 && state.browseURL != "" {
                a.browseToURL(javaPeer, state.browseURL)
                state.browseURL = ""
        }
```

Maybe that receive channel needs to be nil'ed out until `case peer = <-onPeerCreated:` runs?

Or the peer receive case needs to re-send to the updates channel to wake up that case to try again?

Anyway, leaving for you. I haven't actually debugged this other than looking at the code.

**Comment by grantrutherford** [comment]
Same issue on my OnePlus 7 Pro on Android 10. 

I do have the WireGuard app installed but not enabled when testing plus I have the WireGaurd kernel module installed, not sure if this is supported but don't think that's related.

I too have Android System Webview: 83.0.4103.106 installed.

**Comment by frioux** [comment]
Same here:

Device: Pixel 3
OS Version: 10
Tailscale: v0.1.7

**Comment by the1ts** [comment]
Same here

Device Pixel 4XL
OS Version: 10 (June 2020)
Tailscale: v0.1.7

Device Pixel C
OS Version: 8.1 (June 2019)
Tailscale: v0.17

Device Pixel Slate
OS Version: Chrome OS 83.0.4103.112 (android version 9)

**Comment by bradfitz** [comment]
The "Sign In" button works for me (on a vanilla, recently factory reset Pixel 3a), but only if I press it just once and wait 5-10 seconds.

If I press it a bunch of mess with the little "Needs Authentication" widget in the top left corner, then it seems the state machine gets confused and it never works and I have to restart the app to get back to a usable state.

Is anybody able to confirm that it does actually work for you if you're patient? There's a server round-trip involved to get the auth URL and I think the UI just needs to disable the button while that round trip is in flight. (ideally with some spinny/progress animation)

**Comment by bradfitz** [comment]
@eliasnaur, I think there's still more work to be done here, or do you want me to file a new bug for my comment above?

**Comment by maxhawkins** [comment]
Here's a video of me triggering the bug.

https://drive.google.com/file/d/14FZ2RMsSi4-Yu1jSjNDT913yT57Azdxr/view?usp=sharing

And here's the adb logcat:
```
06-19 10:47:17.431 25150 25150 D AppButtonsPrefCtl: Stopping package com.tailscale.ipn
06-19 10:47:17.432  1381  2331 I ActivityManager: Force stopping com.tailscale.ipn appid=10350 user=0: from pid 25150
06-19 10:47:17.433  1381  2331 I ActivityManager: Killing 28605:com.tailscale.ipn/u0a350 (adj 920): stop com.tailscale.ipn
06-19 10:47:17.434  1381  2331 W ActivityTaskManager: Force removing ActivityRecord{fa77c3c u0 com.tailscale.ipn/org.gioui.GioActivity t13945 f}: app died, no saved state
06-19 10:47:17.458 25150 25150 D AppButtonsPrefCtl: Sending broadcast to query restart status for com.tailscale.ipn
06-19 10:47:17.582 25150 25150 D AppButtonsPrefCtl: Sending broadcast to query restart status for com.tailscale.ipn
06-19 10:47:17.589 25150 25150 D AppButtonsPrefCtl: Sending broadcast to query restart status for com.tailscale.ipn
06-19 10:47:17.700 25150 25150 D AppButtonsPrefCtl: Sending broadcast to query restart status for com.tailscale.ipn
06-19 10:47:17.702 25150 25150 D AppButtonsPrefCtl: Got broadcast response: Restart status for com.tailscale.ipn false
06-19 10:47:17.766 25150 25150 D AppButtonsPrefCtl: Got broadcast response: Restart status for com.tailscale.ipn false
06-19 10:47:20.235  1381  4587 I ActivityTaskManager: START u0 {act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10200000 cmp=com.tailscale.ipn/org.gioui.GioActivity bnds=[239,1000][440,1304]} from uid 10042
06-19 10:47:20.313  1381  1473 I ActivityManager: Start proc 31649:com.tailscal.ipn/u0a350 for activity {com.tailscale.ipn/org.gioui.GioActivity}
06-19 10:47:20.323 31649 31649 E m.tailscale.ip: Not starting debugger since process cannot load the jdwp agent.
06-19 10:47:20.401 31649 31649 W m.tailscale.ip: JIT profile information will not be recorded: profile file does not exits.
06-19 10:47:20.401 31649 31649 I chatty  : uid=10350(com.tailscale.ipn) identical 2 lines
06-19 10:47:20.401 31649 31649 W m.tailscale.ip: JIT profile information will not be recorded: profile file does not exits.
06-19 10:47:20.509 31649 31672 I com.tailscale.ipn: 2020/06/19 17:47:20 logtail started
06-19 10:47:20.509 31649 31672 I com.tailscale.ipn: 2020/06/19 17:47:20 Program starting: vdate.20200410, Go devel +3f30e0ee17 Wed Jun 10 21:25:14 2020 +0200: []string{"gojni"}
06-19 10:47:20.509 31649 31672 I com.tailscale.ipn: 2020/06/19 17:47:20 LogID: 8ccf498159afb18419ba0848595f42de8a55c52bf528d95a3ff2c3c82b3fa24e
06-19 10:47:20.583 31649 31649 W m.tailscale.ipn: type=1400 audit(0.0:6799): avc: denied { read } for name="route" dev="proc" ino=4026532078 scontext=u:r:untrusted_app:s0:c94,c257,c512,c768 tcontext=u:object_r:proc_net:s0 tclass=file permissive=0
06-19 10:47:20.629 31649 31682 I com.tailscale.ipn: 2020/06/19 17:47:20 logtail: dialed "log.tailscale.io:443" in 111ms
06-19 10:47:20.703 31649 31649 W m.tailscale.ipn: type=1400 audit(0.0:6800): avc: denied { read } for name="route" dev="proc" ino=4026532078 scontext=u:r:untrusted_app:s0:c94,c257,c512,c768 tcontext=u:object_r:proc_net:s0 tclass=file permissive=0
06-19 10:47:20.703 31649 31649 W m.tailscale.ipn: type=1400 audit(0.0:6801): avc: denied { read } for name="route" dev="proc" ino=4026532078 scontext=u:r:untrusted_app:s0:c94,c257,c512,c768 tcontext=u:object_r:proc_net:s0 tclass=file permissive=0
06-19 10:47:20.716 31649 31649 W m.tailscale.ipn: type=1400 audit(0.0:6802): avc: denied { read } for name="tcp" dev="proc" ino=4026532086 scontext=u:r:untrusted_app:s0:c94,c257,c512,c768 tcontext=u:object_r:proc_net_tcp_udp:s0 tclass=file permissive=0
06-19 10:47:20.804   633   633 E Layer   : [Surface(name=AppWindowToken{6efeeb1 token=Token{5264d58 ActivityRecord{22bf23b u0 com.tailscale.ipn/org.gioui.GioActivity t13946}}})/@0x668e621 - animation-leash#0] No local sync point found
06-19 10:47:20.804   633   633 E Layer   : [Surface(name=AppWindowToken{6efeeb1 token=Token{5264d58 ActivityRecord{22bf23b u0 com.tailscale.ipn/org.gioui.GioActivity t13946}}})/@0x668e621 - animation-leash#0] No local sync point found
06-19 10:47:20.817  1381  1470 I ActivityTaskManager: Displayed com.tailscale.ipn/org.gioui.GioActivity: +577ms
06-19 10:47:51.182  1381  1727 I ActivityTaskManager: START u0 {act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10200000 cmp=com.tailscale.ipn/org.gioui.GioActivity bnds=[239,1000][440,1304]} from uid 10042
06-19 10:47:51.705   633   633 E Layer   : [Surface(name=AppWindowToken{6efeeb1 token=Token{5264d58 ActivityRecord{22bf23b u0 com.tailscale.ipn/org.gioui.GioActivity t13946}}})/@0x668e621 - animation-leash#0] No local sync point found
06-19 10:47:51.705   633   633 E Layer   : [Surface(name=AppWindowToken{6efeeb1 token=Token{5264d58 ActivityRecord{22bf23b u0 com.tailscale.ipn/org.gioui.GioActivity t13946}}})/@0x668e621 - animation-leash#0] No local sync point found
06-19 10:47:51.721   633   633 E Layer   : [Surface(name=AppWindowToken{6efeeb1 token=Token{5264d58 ActivityRecord{22bf23b u0 com.tailscale.ipn/org.gioui.GioActivity t13946}}})/@0x668e621 - animation-leash#0] No local sync point found

```

**Comment by eliasnaur** [comment]
@bradfitz which comment? If you mean the disabling of the button + spinner, then go ahead. I do have an internal list of UI polish tasks once the functionality is stable, but a public GH issue is also good.

**Comment by eliasnaur** [comment]
v0.1.8 is released with the fix for this issue. Let me know if there are still problems.

**Comment by apenwarr** [comment]
Not fixed for me (on ChromeOS) in v0.1.8. If I logout, then click the Login button several times fast, it gets stuck and no login window appears. If I then exit the app, logout again, and click only once, it works.

**Comment by grantrutherford** [comment]
Fixed for me on my OnePlus 7 Pro on Android 10. I now click sign in and it pretty much immediately opens up a window to login, I did have to manually close the window but all good it works. Good work @eliasnaur

**Comment by apenwarr** [comment]
If it "immediately" pops up the browser window for you, that might be the difference: it takes a couple of seconds for me, during which time I have plenty of time to multi-click the button. Not sure why there's a difference.

Sounds like the delay and the "if you click multiple times it gets stuck" were two different underlying bugs, and only one is fixed?

**Comment by grantrutherford** [comment]
I think your bug was slighy different from the bug I had and what's been fixed. For me it was never a time or delay thing, it didn't matter how long I waited I never got the login popup. That's the bug that's been fixed.

Perhaps there is another issue to create here that is something like "You can click on 'sign in' button multiple times which causes the app the lockup".? The fix being it is disabled after the first click and only re-enabled if the action timeout/failed.

**Comment by rosszurowski** [comment]
Got a report of a user on a Google Pixel 3a who can get into the app and go through the authentication flow, but as soon as they finish authenticating, they get kicked out and immediately need to reauthenticate. According to them:

> I get into the UI no problem and "reauthenticating" works just fine.
> But as soon as I go to connect (by clicking the little toggle switch) I get an error and have to "reauthenticate" again.
> A connection is never actually established for any longer than making my device appear on the device list.

Possible it's a separate issue, but it seems this particular one is taking several forms.

**Comment by eliasnaur** [comment]
@apenwarr I capped inflight login URL request to 1. Can you confirm the issue where the login flow locks up is gone on your end?

**Comment by bradfitz** [comment]
@eliasnaur just released v0.1.16 to the beta channel (https://play.google.com/apps/testing/com.tailscale.ipn).

Please let us know if that fixes this for anybody here. Thanks!

**Comment by eliasnaur** [comment]
I'm closing this. Please re-open or file a new issue if the problem(s) persist.
